File: src/justitone/jidi/Periods.java
package justitone.jidi;

import java.util.ArrayList;
import java.util.List;

import org.apache.commons.math3.fraction.BigFraction;
import org.apache.commons.math3.util.Pair;

public class Periods {
    List<Pair<BigFraction, BigFraction>> periods;
    
    public Periods() {
        periods = new ArrayList<>();
    }
    
    public void allocate(BigFraction start, BigFraction end) {
        periods.add(new Pair<>(start, end));
    }
    
    public boolean canAllocate(BigFraction start, BigFraction end) {
        if (periods.size() == 0) return true;
        
        Pair<BigFraction, BigFraction> period = new Pair<>(start, end);
        
        return periods.stream()
                      .allMatch(p -> !overlaps(period, p));
    }
    
    public boolean overlaps(Pair<BigFraction, BigFraction> period1, Pair<BigFraction, BigFraction> period2) {
        if (period1.getFirst().compareTo(period2.getSecond()) >= 0) {
            return false;
        }
        if (period2.getFirst().compareTo(period1.getSecond()) >= 0) {
            return false;
        }
        return true;
    }
}

File: src/main/kotlin/justitone/jidi/Periods.kt
package justitone.jidi

import org.apache.commons.math3.fraction.BigFraction

class Periods {
    private var periods: MutableList<ClosedRange<BigFraction>> = mutableListOf()

    fun allocate(start: BigFraction, end: BigFraction) {
        periods.add(start..end)
    }

    fun canAllocate(start: BigFraction, end: BigFraction): Boolean {
        if (periods.size == 0) return true

        val period = start..end

        return periods.none { !overlaps(period, it) }
    }

    private fun overlaps(period1: ClosedRange<BigFraction>, period2: ClosedRange<BigFraction>): Boolean {
        return if (period1.start >= period2.endInclusive) true
        else period2.start >= period1.endInclusive
    }
}
